### PR TITLE
Pseudo ephemeral anchors

### PR DESCRIPTION
**Pull request by instagibbs**
Accomplishes part of https://github.com/bitcoin-inquisition/bitcoin/issues/12

via first parts of https://github.com/bitcoin/bitcoin/pull/26403

It *does not* require these new output types to be cleaned up via package relay. That is the rest of the logic left out for future updates. It also does nothing to stop pinning attacks; I rely on the V3 nature of the anchors for that.

Added a couple simple tests and split up the logic more than parent PR.

This along with https://github.com/bitcoin-inquisition/bitcoin/pull/22 would allow my ln-symmetry BOLT transactions to enter the mempool with appropriate fees attached, and I could shortly have examples running in my CLN branch with unilateral closes being handled as appropriate.

### CODE REVIEW

File: src/policy/policy.cpp [hunk 1]
```diff
@@ -142,8 +142,12 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
             reason = "bare-multisig";
             return false;
         } else if (IsDust(txout, dust_relay_fee)) {
-            reason = "dust";
-            return false;
+            // TODO: Ensure ephemeral output is spent in the same mempool package
+            // to avoid bloating utxo set
+            if (whichType != TxoutType::ANCHOR) {
```
Comment (naumenkogs) [inline src/policy/policy.cpp]: Do you want to allow *non-dust* outputs to be anchor?

Comment (instagibbs) [reply src/policy/policy.cpp]: Yes that's the idea. This allows protocols to create outputs to dump excess value, instead of dumping it to fees(which complicates future V3/eviction policies).

Comment (naumenkogs) [reply src/policy/policy.cpp]: Can you elaborate, or send me a link if it's already summarized somewhere online?

Comment (instagibbs) [reply src/policy/policy.cpp]: Hmm, looks like I don't discuss it in the bip draft: https://github.com/instagibbs/bips/blob/ephemeral_anchor/bip-ephemeralanchors.mediawiki

Basically, the design choices should be either:
1) it must be 0-value
2) it can be any value

Choosing (2) allows any value that would have gone to "trimmed outputs" in things like LN to go to those outputs, instead of to the commitment transaction fee, since the EA requirement of 0-fee EA transaction exists to make eviction reasoning easier.

See https://github.com/instagibbs/bolts/blob/eltoo_draft/XX-eltoo-transactions.md#ephemeral-anchor-output for an example usage.

Comment (ajtowns) [reply src/policy/policy.cpp]: switch the ordering? `else if (whichType != ANCHOR && IsDust())`

Comment (instagibbs) [reply src/policy/policy.cpp]: reworked @ajtowns

**Comment by instagibbs** [comment]
switched to OP_TRUE as upstream PR/BIP have

File: test/functional/rpc_psbt.py [hunk 2]
```diff
@@ -853,5 +854,20 @@ def test_psbt_input_keys(psbt_input, keys):
         assert_equal(self.nodes[0].combinepsbt([psbt1, psbt1]), psbt1)
 
 
+        self.log.info("Test that PSBT can have ephemeral anchor added in rpc")
+        # Fake input to avoid tripping up segwit deserialization error
+        raw_anchor = self.nodes[0].createrawtransaction([{"txid": "ff"*32, "vout": 0}], [{"anchor":"0.00000001"}])
+        anchor_tx = tx_from_hex(raw_anchor)
+
+        assert_equal(anchor_tx.nVersion, 3)
```
Comment (instagibbs) [inline test/functional/rpc_psbt.py]: guess this doesn't make sense for pseudo-ea huh? Will leave comment and remove

**Comment by ajtowns** [comment]
> It _does not_ require these new output types to be cleaned up via package relay.

Since the point of this is to allow you to have a 0-fee transaction paid for by CPFP, it might be good to have a special exception to `minrelayfee=1000` to make that work while we're still working on package relay. How about something like this:

```diff
diff --git a/src/txmempool.cpp b/src/txmempool.cpp
index b0047c74b2..d0a7fba9b9 100644
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -457,6 +457,19 @@ void CTxMemPool::AddTransactionsUpdated(unsigned int n)
     nTransactionsUpdated += n;
 }
 
+static const CFeeRate EPHEMERAL_DELTA{1000};
+static void ApplyEphemeralDelta(const CTxMemPoolEntry& entry, CAmount& delta)
+{
+    if (delta != 0) return; // overridden?
+    if (entry.GetFee() > 0) return; // only adjust for 0 fee
+    for (const auto& out : entry.GetTx().vout) {
+        if (out.scriptPubKey.IsTrue()) {
+            delta = EPHEMERAL_DELTA.GetFee(entry.GetTxSize());
+            return;
+        }
+    }
+}
+
 void CTxMemPool::addUnchecked(const CTxMemPoolEntry &entry, setEntries &setAncestors, bool validFeeEstimate)
 {
     // Add to memory pool without checking anything.
@@ -467,6 +480,7 @@ void CTxMemPool::addUnchecked(const CTxMemPoolEntry &entry, setEntries &setAnces
     // Update transaction for any feeDelta created by PrioritiseTransaction
     CAmount delta{0};
     ApplyDelta(entry.GetTx().GetHash(), delta);
+    ApplyEphemeralDelta(entry, delta);
     // The following call to UpdateModifiedFee assumes no previous fee modifications
     Assume(entry.GetFee() == entry.GetModifiedFee());
     if (delta) {
```

**Comment by instagibbs** [comment]
@ajtowns took the idea, fixed it up, and put some light test coverage on it. Wallets can now fund a transaction with an ephemeral anchor and put it into the mempool.

**Comment by instagibbs** [comment]
upstream fix for crash here https://github.com/bitcoin/bitcoin/pull/27271

File: src/rpc/rawtransaction_util.cpp [hunk 3]
```diff
@@ -121,7 +121,8 @@ CMutableTransaction ConstructTransaction(const UniValue& inputs_in, const UniVal
                 throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, duplicate key: anchor");
             }
             has_anchor = true;
-            rawTx.nVersion = 3; // Implicitly opts into V3
+            // V3 Transactions are not standard yet
+            // rawTx.nVersion = 3; // Implicitly opts into V3
```
Comment (ajtowns) [inline src/rpc/rawtransaction_util.cpp]: Think it would be better to just not add this line in the first place?

File: src/script/script.cpp [hunk 4]
```diff
@@ -198,13 +198,19 @@ unsigned int CScript::GetSigOpCount(const CScript& scriptSig) const
     return subscript.GetSigOpCount(true);
 }
 
+bool CScript::IsTrue() const
```
Comment (ajtowns) [inline src/script/script.cpp]: I don't really like this function name, though I'm not sure exactly why. Maybe `IsPayToOpTrue` or `IsEphemeralAnchorScript` would be clearer?

File: src/policy/policy.cpp [hunk 5]
```diff
@@ -141,7 +141,7 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
         else if ((whichType == TxoutType::MULTISIG) && (!permit_bare_multisig)) {
             reason = "bare-multisig";
             return false;
-        } else if (whichType != TxoutType::ANCHOR && IsDust(txout, dust_relay_fee)) {
+        } else if ((!permit_ephemeral_anchor || whichType != TxoutType::ANCHOR) && IsDust(txout, dust_relay_fee)) {
```
Comment (ajtowns) [inline src/policy/policy.cpp]: Shouldn't this be:

```c++
} else if (whichType != ANCHOR && IsDust) {
    reason = "dust"; return false;
} else if (!permit_ephemeral_anchor && whichType == ANCHOR) {
    reason = "ephemeral-anchor"; return false;
}
```

(otherwise, even if you set `-ephermalanchors=0` you permit low-feerate relay/mining of large txs if they set their fee to 0 and include a non-dust OP_TRUE output)

File: src/rpc/rawtransaction.cpp [hunk 6]
```diff
@@ -154,6 +154,11 @@ static std::vector<RPCArg> CreateTxDoc()
                         {"data", RPCArg::Type::STR_HEX, RPCArg::Optional::NO, "A key-value pair. The key must be \"data\", the value is hex-encoded data"},
                     },
                 },
+                {"", RPCArg::Type::OBJ, RPCArg::Optional::OMITTED, "",
+                    {
+                        {"anchor", RPCArg::Type::STR_HEX, RPCArg::Optional::NO, "Creates an ephemeral anchor. A key-value pair. The key must be \"anchor\", the value is the amount in " + CURRENCY_UNIT},
```
Comment (ajtowns) [inline src/rpc/rawtransaction.cpp]: If the value is an amount, it should be `Type::AMOUNT` not `Type::STR_HEX` presumably

File: src/txmempool.cpp [hunk 7]
```diff
@@ -457,6 +457,22 @@ void CTxMemPool::AddTransactionsUpdated(unsigned int n)
     nTransactionsUpdated += n;
 }
 
+/* Inquisition cheat to get around lack of package relay. Will only work
+ * with default minrelayfee, with non-full mempools.
+ */
+static const CFeeRate EPHEMERAL_DELTA{1000};
```
Comment (ajtowns) [inline src/txmempool.cpp]: Maybe this value should be configurable? Then you could say `-ephemeraldelta=0.0000005 -minrelaytxfee=0.0000005 -blockmintxfee=0.00001`  so that you'd only mine txs with ephemeral anchors if they were CPFPed to 1s/vb.

File: src/wallet/spend.cpp [hunk 8]
```diff
@@ -899,6 +899,13 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
     const CAmount not_input_fees = coin_selection_params.m_effective_feerate.GetFee(coin_selection_params.tx_noinputs_size);
     CAmount selection_target = recipients_sum + not_input_fees;
 
+    // This can only happen if feerate is 0, and requested destinations are value of 0 (i.e. ephemeral anchor)
```
Comment (ajtowns) [inline src/wallet/spend.cpp]: Having this hunk be in its own cherry-picked commit might make merging easier if bitcoin/bitcoin#27271 gets backported?

File: src/script/script.cpp (hunk 4 above)
Comment (instagibbs) [reply src/script/script.cpp]: `IsPayToAnchor`? That's what I switched to to better match current naming schemes

File: src/txmempool.cpp (hunk 7 above)
Comment (instagibbs) [reply src/txmempool.cpp]: that would require everyone on the path to set the first two options to propagate, or signet miner to a higher blockmintxfee?

File: src/wallet/spend.cpp [hunk 9]
```diff
@@ -889,7 +889,7 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
             coin_selection_params.tx_noinputs_size += ::GetSerializeSize(txout, PROTOCOL_VERSION);
         }
 
-        if (IsDust(txout, wallet.chain().relayDustFee())) {
+        if (!recipient.scriptPubKey.IsTrue() && IsDust(txout, wallet.chain().relayDustFee())) {
```
Comment (instagibbs) [inline src/wallet/spend.cpp]: probably shouldn't allow making a tx you will not propagate due to mempool flags

File: src/rpc/rawtransaction_util.cpp (hunk 3 above)
Comment (instagibbs) [reply src/rpc/rawtransaction_util.cpp]: done

File: src/policy/policy.cpp (hunk 5 above)
Comment (instagibbs) [reply src/policy/policy.cpp]: done

File: src/wallet/spend.cpp (hunk 8 above)
Comment (instagibbs) [reply src/wallet/spend.cpp]: done

File: src/rpc/rawtransaction.cpp (hunk 6 above)
Comment (instagibbs) [reply src/rpc/rawtransaction.cpp]: done

File: src/script/script.cpp (hunk 4 above)
Comment (ajtowns) [reply src/script/script.cpp]: Sounds good

File: src/txmempool.cpp (hunk 7 above)
Comment (ajtowns) [reply src/txmempool.cpp]: I was thinking miner would override as above, everyone else would just have them all as 1000sat/kvb -- that way everything gets relayed, but it only gets mined if there's a real-ish cpfp.

Comment (ajtowns) [reply src/txmempool.cpp]: (Could leave that for another PR though)

Comment (instagibbs) [reply src/txmempool.cpp]: pushed a commit which adds `ephemeraldelta` with test

**Comment by instagibbs** [comment]
Added another commit from master which fixes PSBT finalization of ephemeral anchors: https://github.com/bitcoin/bitcoin/pull/25595/commits and a simple test showing how to use Core's wallet to bump anchors manually.

**Comment by ajtowns** [comment]
Needs rebase

**Comment by instagibbs** [comment]
rebased

File: src/policy/policy.cpp [hunk 10]
```diff
@@ -141,9 +141,12 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
         else if ((whichType == TxoutType::MULTISIG) && (!permit_bare_multisig)) {
             reason = "bare-multisig";
             return false;
-        } else if (IsDust(txout, dust_relay_fee)) {
+        } else if (whichType != TxoutType::ANCHOR && IsDust(txout, dust_relay_fee)) {
             reason = "dust";
             return false;
+        } else if (whichType == TxoutType::ANCHOR && !permit_ephemeral_anchor) {
+            reason = "ephemeral-anchor"; return false;
+            return false;
```
Comment (ajtowns) [inline src/policy/policy.cpp:151]: Should there be another check for `multiple-ephemeral-anchor` ? (If not, why the "duplicate key: anchor" check in rpc?)

**Review by ajtowns — COMMENTED** [review]
LGTM

File: src/policy/policy.cpp (hunk 10 above)
Comment (instagibbs) [reply src/policy/policy.cpp:151]: there's no actual use case for 2+, so this would neatly apply onto my branch against master which *does* restrict it via policy. I could port more logic over to make these checks, of course.

Comment (ajtowns) [reply src/policy/policy.cpp:151]: I guess I was thinking that `CountEphmeralOutputs` should just be `bool HasEphemeralOutput` and the `too-many-ephmeral-anchors` check should be here, same as `multi-op-return`.

File: src/wallet/spend.cpp [hunk 11]
```diff
@@ -889,16 +889,34 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
...
     }
 
     // Include the fees for things that aren't inputs, excluding the change output
     const CAmount not_input_fees = coin_selection_params.m_effective_feerate.GetFee(coin_selection_params.tx_noinputs_size);
     CAmount selection_target = recipients_sum + not_input_fees;
 
+    // This can only happen if feerate is 0, and requested destinations are value of 0 (i.e. ephemeral anchor)
+    // and no pre-selected inputs. This will result in 0-input transaction, which we cannot support without
+    // modifying coin selection algorithm for such a niche use-case.
+    if (selection_target == 0 && !coin_control.HasSelected()) {
```
Comment (instagibbs) [inline src/wallet/spend.cpp]: "I'm seeing double: four krustys!"

**Comment by instagibbs** [comment]
Pushed updates, including 0-fee check for ephemeral anchor transactions. 

Also pushed https://github.com/bitcoin-inquisition/bitcoin/pull/23/commits/7e10a460d3686e4c010594d77efaded9e4404072 which may or may not be a good idea, but makes sure the caller explicitly knows they have to get that value from somewhere when making `IsDust` determinations. Seems less brittle?

File: src/policy/policy.cpp (hunk 10 above)
Comment (instagibbs) [reply src/policy/policy.cpp:151]: done. also added 0-fee check in `PreChecks`

**Comment by instagibbs** [comment]
From here to "full EA" we basically just have:

1) check for v3 if anchor is found(could happen in IsStandardTx now)
2) V3-related package checks making sure it's spent
3) tests

File: src/policy/policy.cpp [hunk 12]
```diff
@@ -62,8 +62,10 @@ CAmount GetDustThreshold(const CTxOut& txout, const CFeeRate& dustRelayFeeIn)
     return dustRelayFeeIn.GetFee(nSize);
 }
 
-bool IsDust(const CTxOut& txout, const CFeeRate& dustRelayFeeIn)
+bool IsDust(const CTxOut& txout, const CFeeRate& dustRelayFeeIn, bool permit_dust_anchor)
 {
+    if (permit_dust_anchor && txout.scriptPubKey.IsPayToAnchor()) return false;
```
Comment (ajtowns) [inline src/policy/policy.cpp]: I think this would be better if you just did `if (txout.sPK.IsPayToAnchor()) return 0;` in `GetDustThreshold()`, and moved the `} else if (whichType == ANCHOR && !permit_anchor) {` check to be prior to the dust one? (Then you don't need the extra parameter for `IsDust()`)

I don't think there's much point special casing "non-dust OP_TRUE output" as being okay if you disable ephemeral anchors; old nodes still won't relay it, and future nodes will be doing v3 package relay anyway.

Comment (instagibbs) [reply src/policy/policy.cpp]: Well it's more if someone turns off the feature, they may still think something is not dust based on various interfaces e.g., GUI that don't use `IsStandardTx`. I'm all for simpler changes of course...

Comment (ajtowns) [reply src/policy/policy.cpp]: It's a signet only feature that's not even targeted at mainnet ("pseudo"), and doesn't have a use case ("i want to send meaningful amounts of funds to whoever mines this tx, but in a more complicated way than dropping it to fees")... Yeah, people could think weird edge cases are okay, but...

Comment (instagibbs) [reply src/policy/policy.cpp]: I'm attempting to minimize the diff of this PR from the one that could go towards Core master, and thinking aloud. I think going with simpler is the right default choice; complexity can be tacked on later.

I suppose we somehow survive this issue with OP_RETURN's. Good enough for me!

**Comment by ajtowns** [comment]
Lots of missing test coverage? Could add non-zero-fee test via data/invalid_txs.py at least?

Is there any reason for bitcoin core to care if a tx with an ephemeral anchor output is version 3? Isn't it sufficient to allow ephemeral anchors anywhere, and just rely on protocol designers to make it v3 presuming they want pinning protection? (If so, that might let you drop the version 3 bits from the functional test, so that you could add it here?)

**Comment by instagibbs** [comment]
The "hardest part" is making sure the utxo is spent within the package. These checks are embedded in V3-only checks, since the topology becomes exceedingly simple. We can continue litigating that decision, but that's a bit out of scope since that logic doesn't exist here.

re:tests, I just pushed 4 of the 6 test cases, removing the test cases for making sure EA are spent, and package RBF cases, which require package RBF.

**Comment by ajtowns** [comment]
ACK 296adcfb0691f2f73325ac8852deb75125a88583

File: src/policy/policy.h [hunk 13]
```diff
@@ -136,6 +140,11 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
 */
 bool IsWitnessStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs, bool allow_annex_data);
 
+/**
+* Check if given transaction has a IsPayToAnchor output.
+**/
+size_t HasEphemeralAnchor(const CTransaction& tx);
```
Comment (kallewoof) [inline src/policy/policy.h:146]: It seems like `IsEphemeralAnchor` / `HasEphemeralAnchor`, or `IsPayToAnchor` / `HasPayToAnchor` would be more intuitive names. These sound like different things here.

File: src/script/standard.cpp [hunk 14]
```diff
@@ -52,6 +52,7 @@ std::string GetTxnOutputType(TxoutType t)
     case TxoutType::SCRIPTHASH: return "scripthash";
     case TxoutType::MULTISIG: return "multisig";
     case TxoutType::NULL_DATA: return "nulldata";
+    case TxoutType::ANCHOR: return "true";
```
Comment (kallewoof) [inline src/script/standard.cpp:55]: This looks weird. I understand that it's meant to say "anyone can spend". The type could be named "`TxnoutType::TRUE`" here, and be used by the anchor stuff. Or this could return "`anchor`". Either way would be better, I think.

File: test/functional/feature_segwit.py [hunk 15]
```diff
@@ -448,18 +448,18 @@ def run_test(self):
...
             unsolvablep2pkh = key_to_p2pkh_script(unsolvable_address_key)
             unsolvablep2wshp2pkh = script_to_p2wsh_script(unsolvablep2pkh)
             p2shop0 = script_to_p2sh_script(op0)
-            p2wshop1 = script_to_p2wsh_script(op1)
+            p2wshop2 = script_to_p2wsh_script(op2)
             unsolvable_after_importaddress.append(unsolvablep2pkh)
             unsolvable_after_importaddress.append(unsolvablep2wshp2pkh)
-            unsolvable_after_importaddress.append(op1)  # OP_1 will be imported as script
-            unsolvable_after_importaddress.append(p2wshop1)
+            unsolvable_after_importaddress.append(op2)  # OP_1 will be imported as script
```
Comment (kallewoof) [inline test/functional/feature_segwit.py:461]: Comment still says `OP_1`.

File: test/functional/mempool_ephemeral_anchor.py [hunk 16]
```diff
@@ -0,0 +1,211 @@
+#!/usr/bin/env python3
+# Copyright (c) 2022 The Bitcoin Core developers
```
Comment (kallewoof) [inline test/functional/mempool_ephemeral_anchor.py:2]: 2023 now.

File: test/functional/test_framework/wallet.py [hunk 17]
```diff
@@ -67,18 +68,18 @@ class MiniWalletMode(Enum):
...
-    should contain an actual signature, the raw modes RAW_OP_TRUE and RAW_P2PK
+    should contain an actual signature, the raw modes RAW_OP_2 and RAW_P2PK
     can be useful. Summary of modes:
 
                     |      output       |           |  tx is   | can modify |  needs
          mode       |    description    |  address  | standard | scriptSig  | signing
     ----------------+-------------------+-----------+----------+------------+----------
     ADDRESS_OP_TRUE | anyone-can-spend  |  bech32m  |   yes    |    no      |   no
-    RAW_OP_TRUE     | anyone-can-spend  |  - (raw)  |   no     |    yes     |   no
+    RAW_OP_2     | anyone-can-spend  |  - (raw)  |   no     |    yes     |   no
```
Comment (kallewoof) [inline test/functional/test_framework/wallet.py:78]: Spacing got borked.

**Review by kallewoof — COMMENTED** [review]
Post-merge ACK

File: src/script/standard.cpp (hunk 14 above)
Comment (instagibbs) [reply src/script/standard.cpp:55]: >  The type could be named "TxnoutType::TRUE

I don't think I can pick an enum `TRUE` as it's a reserved word in windows.

Could be "anchor" here sure.

**Comment by instagibbs** [comment]
thanks @kallewoof , took these changes on my local branch for future rebasing/use upstream

File: src/script/standard.cpp (hunk 14 above)
Comment (instagibbs) [reply src/script/standard.cpp:55]: switching to "anchor" everywhere

**Comment by ajtowns** [comment]
> thanks @kallewoof , took these changes on my local branch for future rebasing/use upstream

Any reason not to add them as a new inquisition PR as well?

**Comment by instagibbs** [comment]
https://github.com/bitcoin-inquisition/bitcoin/pull/26 done

**Comment by instagibbs** [comment]
hmmm, hole in this theory is that p2p will not propagate txns below feefilter rates advertised by the peer. Can be manually submitted over RPC

**Comment by ajtowns** [comment]
> hmmm, hole in this theory is that p2p will not propagate txns below feefilter rates advertised by the peer. Can be manually submitted over RPC

I think this works fine once you spend the ephemeral anchor -- the tx that spends it passes feefilter so does get announced, then the tx with the EA output gets requested via orphan processing.

**Comment by levantah** [comment]
Just a note that this PR is no more part of `v27.0-inq` per https://delvingbitcoin.org/t/bitcoin-inquisition-27-0/883:

> This no longer includes relay support for annexdatacarrier (see https://github.com/bitcoin-inquisition/bitcoin/pull/22) or pseudo ephemeral anchors (see https://github.com/bitcoin-inquisition/bitcoin/pull/23). The signet inquisition miner and inquisition relay node mentioned above will likely be updated in the next week or two, at which point transactions using these features will no longer confirm.
